sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java

```diff
@@ -165,17 +147,59 @@ public MetricData collectAndReset(
           instrumentationLibraryInfo,
           getMetricDescriptor(),
           temporality,
-          asyncAccumulator.collectAndReset(),
+          accumulator.collectAndReset(),
           startEpochNanos,
           epochNanos);
     } finally {
       collectLock.unlock();
     }
   }
 
-  @Override
-  public MetricDescriptor getMetricDescriptor() {
-    return metricDescriptor;
+  /**
+   * Add a callback to the storage. Once all callbacks have already been removed via {@link
+   * #removeCallback(Consumer)}, subsequent registrations will fail.
+   */
+  public void addCallback(Consumer<O> callback) {
+    callbackLock.lock();
+    try {
+      if (unregistered) {
+        throttlingLogger.log(
+            Level.WARNING,
+            "Attempting to add callback for "
+                + getMetricDescriptor().getName()
+                + " after storage has been unregistered.");
+        return;
+      }
+      this.callbacks.add(callback);
+    } finally {
+      callbackLock.unlock();
+    }
+  }
+
+  /**
+   * Remove the callback from the storage. Called when {@link ObservableInstrument#remove()} is
+   * invoked. When the last callback is removed, the storage is unregistered from {@link
+   * #metricStorageRegistry}.
+   */
+  public void removeCallback(Consumer<O> callback) {
+    callbackLock.lock();
+    try {
+      if (unregistered) {
+        throttlingLogger.log(
+            Level.WARNING,
+            "Attempting to remove callback for "
+                + getMetricDescriptor().getName()
+                + " after storage has been unregistered.");
+        return;
+      }
+      this.callbacks.remove(callback);
+      if (callbacks.size() == 0) {
+        metricStorageRegistry.unregister(this);
```

Comment:
It doesn't seem like removing a callback should cause total unregistration. From my reading of the code, it looks like this should but wouldn't work

```
addCallback
removeCallback
addCallback
```

Second `addCallback` would be `unregistered`

If the storage gets recreated at a higher level then maybe it's ok but I'd expect symmetry either at this level or a higher level, currently add and remove aren't symmetric and makes me wonder if there is possible thread safety issues at that higher level if this is true.

Reply:
Also I wonder if unregistration causes the metric to stop being reported (not sure so just confirming) - if that is true, then we should avoid it for another reason, we'd want async and sync behavior to be consistent. There is currently no way for a sync metric to stop being reported, so we would want to solve both at the same time rather than just one.

Reply:
> If the storage gets recreated at a higher level then maybe it's ok 

That's what happening - the second time you call `addCallback` a new `AynchronousMetricStorage` is registered with the `MetricStorageRegistry` because none currently exists at that point. Its likely an unnecessary optimization, since it only frees up resources of asynchronous metrics that have no more callbacks registered at all, which is probably unusual. 

> Also I wonder if unregistration causes the metric to stop being reported (not sure so just confirming) 

Asynchronous metrics are different from synchronous metrics in that we delegate the management of the metrics stream to the callback. So today (i.e. without this code) if a callback stops reporting a metric stream, it will stop being exported to cumulative exporters. In contrast, synchronous instruments continue to report values to exporters even when there is no data reported to the metric stream during the collection interval. This PR doesn't address that behavior. 

Here's a little demo of what I mean:
```
    InMemoryMetricReader cumulativeReader = InMemoryMetricReader.create();
    Meter meter = SdkMeterProvider.builder()
        .registerMetricReader(cumulativeReader)
        .setMinimumCollectionInterval(Duration.ofSeconds(0))
        .build()
        .get("test-meter");

    AtomicLong counter = new AtomicLong(2);
    meter.counterBuilder("foo").buildWithCallback(measurement -> {
      long count = counter.decrementAndGet();
      if (count >= 0) {
        measurement.record(10);
      }
    });

    for (int i = 0; i < 5; i++) {
      System.out.println(cumulativeReader.collectAllMetrics());
    }
```
That code yields:
```
[...data...]
[...data...]
[]
[]
[]
```